### ISSUE TITLE
Group per screen

### ISSUE THREAD

**Issue by AeroNotix** [stumpwm/stumpwm#104]
Hi,

Is there a way to get groups per screen?

At the moment when I have multiple monitors and I change groups, then both screens move to the new group (looks like the group is extended across both monitors).

Is there a way to get it so that each screen has a different group (or a hack to make it seem so) ?

**Comment by grouzen** [comment]
Just want to say that I have the same problem as AeroNotix had.
On both WMs(xmonad, stumpwm) I run 'xrandr --output VGA1 --mode 1920x1080 --right-of LVDS1', and that's what I've got:
http://dump.bitcheese.net/images/ypifyce/2015-01-31-024533_3286x1080_scrot.png - xmonad
http://dump.bitcheese.net/images/etimepy/2015-01-31-024441_3286x1080_scrot.png - stumpwm

**Comment by jorams** [comment]
@grouzen I don't think that's the same problem AeroNotix had. His issue is basically that StumpWM groups, the closest equivalent to xmonad's (and many other wm's) workspaces, are for the entire group of "heads", instead of per "head". Your problem seems to be a bug somewhere, not a design choice.

**Comment by grouzen** [comment]
@jorams, Yeah, it's probably a bug, or even some issue in arhitecture of stumpwm, because when I'm getting the list of screens from stumpwm:*screen-list* variable, I see only one screen instead of two. I think that's the problem. The next strange bug is about geometry of that screen. I connect my laptop to the big monitor(1920x1080), get *screen-list* and I see that geometry of screen object isn't changed - it's still 1366x768. I'm not sure yet if stumpwm is watching for changes in screens configuration on the fly, maybe this is the root of cause of this bug.

**Comment by jorams** [comment]
@grouzen That's an issue of terminology. "Screens" as used by StumpWM (and X) are not monitors. The term used for monitors (what we humans tend to call screens) is "head".

**Comment by dbjergaard** [comment]
@grouzen, it sounds like you need the interactive command
'refresh-heads'.  The symptoms your describing appear if you are using a
settings daemon that is hijacking the configure-notify request from
stumpwm.  Refresh-heads forces stumpwm to re-evaluate the head and
screen structures.  Its not perfect, but it gets the job done.

**Comment by grouzen** [comment]
@jorams, yeah, I understand this, but 'head' is a term of xinerama in my meaning, I'm saying about 'screens' in termnilogy of X11.

**Comment by grouzen** [comment]
@dbjegaard, I see, thanks. I will try 'refresh-heads' when I'll be able to do this next time.

**Comment by AeroNotix** [comment]
I must say though, despite the thing being written in Lisp, there's only a single supported "workflow". It seems it runs in opposition to the "hacker" ethos.

Would people be fine with the ability to do both?

What work would be involved in changing stumpwm to support both?

**Comment by dbjergaard** [comment]
@grouzen sorry about the arcane terminology, stumpwm grew up at a time when Xinerama was still evolving.  This means that stump has certain xinerama concepts even though Xinerama is no longer around.

@AeroNotix Both what? StumpWM is hackable, but it is infeasible to add and support every workflow.  StumpWM borrows concepts from emacs and GNU screen as a reasonable default.  Beyond that its on the user to customize it to their liking.  If you prefer a different workflow and stumpwm doesn't have the hooks you need, then add them. If your changes fit within the module system, then write it as a module.  If you need hooks accessible to StumpWM internals, add them and utilize them from your module.

In this specific instance: I suspect (though I'm often wrong) that StumpWM's representation of groups is associated at the screen level.  What you are requesting is groups at the head level.  What it would take to change the two I don't have a clue. If you can come up with a patch that allows the user to toggle an option to have head based groups vs screen based groups, then please submit the PR!

**Comment by AeroNotix** [comment]
@dbjergaard that's my question. I don't know where to start with that, pointers appreciated! @jorams would you know?

**Comment by dbjergaard** [comment]
I would start with add-group in group.lisp and work forwards and backwards from there. Sorry I can't give any more hints.  You'll have to read the source and build an idea of how the parts fit together in order to implement what you want.

**Comment by jorams** [comment]
@AeroNotix I'm not sure, really. A good place to start could be to modify screens so that they create a group for every head. Then see what goes wrong (which is probably a lot). You might need to look at all invocations of `screen-current-group` to see what they do with that, because they might cause trouble.

I suspect you'll have the least trouble from the floating group, because it pretty much ignores everything to do with heads.

---
Rant:

There's all kinds of things in StumpWM that I wish were implemented in a more general way. There's an incredibly low amount of abstraction, meaning all changes can have far more consequences than they should. Lots of functions (and variables etc.) are defined in places that seem mostly unrelated. There's a group API that allows different types of groups to be defined, but the codebase really isn't written with flexibility in mind. It kind of doesn't feel like a Lisp project in that respect.

During the past few months I have, on more than one occasion, considered building a new window manager from scratch, but I don't really want to do that. At the same time converting StumpWM to be the "perfect" window manager seems almost impossible. In day to day usage it's pretty much good enough for me, so that doesn't serve as too strong a motivator either.

/rant.

**Comment by dbjergaard** [comment]
@jorams, I feel your pain.  Unfortunately this is a side-project for
many of us, and I personally don't have the time or expertise to make
StumpWM live up to its promises. Without reading the source at all,
guile-wm may be structured better since it bill itself as a "A Window
Manger Toolkit for Guile Scheme", with an emacs-like interface
implemented by default.  Unfortunately that project is much less mature
than StumpWM's codebase.

Part of the issue is that StumpWM has its feet in too worlds: It was
written as a re-implementation of ratpoison in lisp.  Unfortunately this
means that a lot of c-like idioms make it into the way stumpwm is
structured. If there are obvious places that could benefit from
refactoring via abstraction, I would be very interested in a discussion
for implementing it.

**Comment by AeroNotix** [comment]
@jorams Perhaps it'd be nice to go through what you think is lacking and we can see whether or not it's feasible to improve that? Providing it's easy to quantify.

**Comment by grouzen** [comment]
@dbjergaard, I've tried the 'refresh-heads' command - unfortunately it doesn't work, nothing have changed.

**Comment by jorams** [comment]
@grouzen Random guess, do you have `xdpyinfo` installed?

**Comment by grouzen** [comment]
@jorams Yeah, I have. When I connect external monitor, I still have one screen(xpdyinfo output) but with resolution of the external monitor.

**Comment by dbjergaard** [comment]
@grouzen Since you're dealing with a true bug, can you please open another issue and detail the steps to reproduce, and any relevent system information (distro, sbcl version etc).  https://github.com/stumpwm/stumpwm/wiki/Bugs

**Comment by grouzen** [comment]
@dbjergaard Yes, sure!

**Comment by mhaffner** [comment]
Has there been any progress on this (or any desire to implement it in the future)?

**Comment by AeroNotix** [comment]
At the time (four year's ago!) I attempted to implement this and discussed a lot with @jorams via IRC various changes necessary. His concerns about the codebase at the time were entirely justified and it was a chore to make changes so unfortunately I ultimately, got bored. 

That said, I've since moved to a single monitor. I'll probably start using stump again. This issue is the single reason I stopped using it.
